services/pagerduty2/service.go

```diff
@@ -0,0 +1,378 @@
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Accept", "application/vnd.pagerduty+json;version=2")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusAccepted && resp.StatusCode != http.StatusOK {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			fmt.Printf("error parsing error body\n")
+			return err
+		}
+		type response struct {
+			Message string `json:"message"`
+		}
+		r := &response{Message: fmt.Sprintf("failed to understand PagerDuty2 response. code: %d content: %s", resp.StatusCode, string(body))}
+		b := bytes.NewReader(body)
+		dec := json.NewDecoder(b)
+		dec.Decode(r)
+		return errors.New(r.Message)
+	}
+	return nil
+}
+
+func (s *Service) sendResolve(c Config, serviceKey, incidentKey string) (string, io.Reader, error) {
+	// create a new AlertPayload for us to fire off
+	type Resolve struct {
+		RoutingKey  string `json:"routing_key"`
+		DedupKey    string `json:"dedup_key"`
+		EventAction string `json:"event_action"`
+	}
+
+	ap := Resolve{}
+
+	if serviceKey == "" {
+		ap.RoutingKey = c.ServiceKey
+	} else {
+		ap.RoutingKey = serviceKey
+	}
+
+	ap.DedupKey = incidentKey
+	ap.EventAction = "resolve"
+
+	var post bytes.Buffer
+	enc := json.NewEncoder(&post)
+	err := enc.Encode(ap)
+	if err != nil {
+		return "", nil, err
+	}
+
+	return c.URL, &post, nil
+}
+
+// preparePost is a helper method that sets up the payload for transmission to PagerDuty
+func (s *Service) preparePost(serviceKey, incidentKey, desc string, level alert.Level, timestamp time.Time, data alert.EventData) (string, io.Reader, error) {
+	c := s.config()
+	if !c.Enabled {
+		return "", nil, errors.New("service is not enabled")
+	}
+
+	var severity string
+	eventType := "trigger"
+
+	switch level {
+	case alert.Warning:
+		severity = "warning"
+	case alert.Critical:
+		severity = "critical"
+	case alert.Info:
+		severity = "info"
+	default:
+		// default is a 'resolve' function
+		return s.sendResolve(c, serviceKey, incidentKey)
+	}
+
+	// create a new AlertPayload for us to fire off
+	ap := &AlertPayload{
+		Payload: &PDCEF{},
+	}
+
+	if serviceKey == "" {
+		ap.RoutingKey = c.ServiceKey
+	} else {
+		ap.RoutingKey = serviceKey
+	}
+
+	ap.Client = "kapacitor"
+	ap.ClientURL = s.HTTPDService.URL()
+	ap.DedupKey = incidentKey
+	ap.EventAction = eventType
+
+	ap.Payload.CustomDetails = make(map[string]interface{})
+	ap.Payload.CustomDetails["result"] = data.Result
+
+	// The API doesn't explicitly mention a requirement for nanosecond resolution but payloads seem to
+	// fail if we don't include it (even zeroes). This hack is not graceful, but adds a negligible
+	// nanosecond resolution to our timestamp
+	m := timestamp.Format(time.RFC3339Nano)
+	if timestamp.Nanosecond() == 0 {
+		m = time.Unix(timestamp.Unix(), 1).In(timestamp.Location()).Format(time.RFC3339Nano)
+	}
```

Comment:
This is just trying to ensure that the timestamps are correctly padded with zeros? And it does so by adding a single nanosecond? If so there is another way to do this. Go's time formatting docs don't make it super clear but in the format string if you use the digit `0` as a placeholder then it will correctly pad the string to be a fixed width.

From the Go source:

```
RFC3339Nano = "2006-01-02T15:04:05.999999999Z07:00"
                            
```
If you use this format instead, then zero padding will be added.

```
"2006-01-02T15:04:05.000000000Z07:00"
```

See https://play.golang.org/p/87-3z6vnUnc